node: e2e: fix device plugin reboot test

ffromani opened the pull request:
#### What type of PR is this?

/kind bug
/kind failing-test

#### What this PR does / why we need it:
Clarifies and fixes misleading e2e device manager test which was permafailing roughly since https://github.com/kubernetes/kubernetes/pull/127506 which also included https://github.com/kubernetes/kubernetes/pull/120661, which has the actual change, was merged.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/128443

#### Special notes for your reviewer:
Replaces https://github.com/kubernetes/kubernetes/pull/128870
Alternative to https://github.com/kubernetes/kubernetes/pull/128995

Reviewer please be aware there's quite some history and context to grok the rationale of this change, which I tried my best to distillate in the commit message, which is unfortunately quite long and verbose for this reason.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

ffromani committed:
e2e: node: fix misleading device plugin test

We have a e2e test which tries to ensure device plugin assignments to pods are kept
across node reboots. And this tests is permafailing since many weeks at
time of writing (xref: #128443).

Problem is: closer inspection reveals the test was well intentioned, but
puzzling:
The test runs a pod, then restarts the kubelet, then _expects the pod to
end up in admission failure_ and yet _ensure the device assignment is
kept_! https://github.com/kubernetes/kubernetes/blob/v1.32.0-rc.0/test/e2e_node/device_plugin_test.go#L97

A reader can legitmately wonder if this means the device will be kept busy forever?

This is not the case, luckily. The test however embodied the behavior at
time of the kubelet, in turn caused by #103979

Device manager used to record the last admitted pod and forcibly added
to the list of active pod. The retention logic had space for exactly one
pod, the last which attempted admission.

This retention prevented the cleanup code
(see: https://github.com/kubernetes/kubernetes/blob/v1.32.0-rc.0/pkg/kubelet/cm/devicemanager/manager.go#L549
compare to: https://github.com/kubernetes/kubernetes/blob/v1.31.0-rc.0/pkg/kubelet/cm/devicemanager/manager.go#L549)
to clear the registration, so the device was still (mis)reported
allocated to the failed pod.

This fact was in turn leveraged by the test in question:
the test uses the podresources API to learn about the device assignment,
and because of the chain of events above the pod failed admission yet
was still reported as owning the device.

What happened however was the next pod trying admission would have
replaced the previous pod in the device manager data, so the previous
pod was no longer forced to be added into the active list, so its
assignment were correctly cleared once the cleanup code runs;
And the cleanup code is run, among other things, every time device
manager is asked to allocated devices and every time podresources API
queries the device assignment

Later, in PR https://github.com/kubernetes/kubernetes/pull/120661
the forced retention logic was removed from all the resource managers,
thus also from device manager, and this is what caused the permafailure.

Because all of the above, it should be evident that the e2e test was
actually enforcing a very specific and not really work-as-intended
behavior, which was also overall quite puzzling for users.

The best we can do is to fix the test to record and ensure that
pods which did fail admission _do not_ retain device assignment.

Unfortunately, we _cannot_ guarantee the desirable property that
pod going running retain their device assignment across node reboots.

In the kubelet restart flow, all pods race to be admitted. There's no
order enforced between device plugin pods and application pods.
Unless an application pod is lucky enough to _lose_ the race with both
the device plugin (to go running before the app pod does) and _also_
with the kubelet (which needs to set devices healthy before the pod
tries admission).

Signed-off-by: Francesco Romani <fromani@redhat.com>

bart0sh commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -934,7 +934,10 @@ func testDevicePluginNodeReboot(f *framework.Framework, pluginSockDir string) {
+		// NOTE: this is flaky because all pods race to go running when kubelet restarts. So application pod will run with device plugion pod.
+		// It's possible, even though unlikely, that the device plugin pod manages to go running *and* the kubelet reports device healthy
+		// when the application pod enters admission. Hence the flakes.
+		framework.It("Does not keeps device plugin assignments across node reboots if fails admission (no pod restart, no device plugin re-registration)", framework.WithFlaky(), func(ctx context.Context) {
```
typo: keeps -> keep

ffromani replied:
thanks, fixed!

bart0sh commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -934,7 +934,10 @@ func testDevicePluginNodeReboot(f *framework.Framework, pluginSockDir string) {
-		framework.It("Keeps device plugin assignments across node reboots (no pod restart, no device plugin re-registration)", framework.WithFlaky(), func(ctx context.Context) {
+		// NOTE: this is flaky because all pods race to go running when kubelet restarts. So application pod will run with device plugion pod.
+		// It's possible, even though unlikely, that the device plugin pod manages to go running *and* the kubelet reports device healthy
+		// when the application pod enters admission. Hence the flakes.
```
Is it possible to postpone admission of the application pod until device plugin pod is registered? It would make the whole setup much more predictable.

ffromani replied:
we will need to add the admission retry loop in kubelet, which we talk about since quite some time (e.g. https://github.com/kubernetes/kubernetes/issues/119248 - note this is not limited to device plugins even though device plugins are a prominent use case) and add some form of dependency tracking between pods; or perhaps, something like scheduling gates (https://kubernetes.io/blog/2022/12/26/pod-scheduling-readiness-alpha/) but for kubelet admission: "don't even try admission until this condition is met; retry later" + a timeout.

likakuli commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -934,7 +934,10 @@ func testDevicePluginNodeReboot(f *framework.Framework, pluginSockDir string) {
 		// intentionally a scenario is created where after node reboot, application pods requesting devices appear before the device plugin pod
 		// exposing those devices as resource has restarted. The expected behavior is that the application pod fails at admission time.
-		framework.It("Keeps device plugin assignments across node reboots (no pod restart, no device plugin re-registration)", framework.WithFlaky(), func(ctx context.Context) {
+		// NOTE: this is flaky because all pods race to go running when kubelet restarts. So application pod will run with device plugion pod.
```
typo: plugion -> plugin

ffromani commented:
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-node-kubelet-containerd-flaky

ffromani commented:
known failure, change on the way

ffromani commented:
/test pull-kubernetes-node-kubelet-containerd-flaky
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-e2e-gce

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -934,7 +934,10 @@ func testDevicePluginNodeReboot(f *framework.Framework, pluginSockDir string) {
 		// intentionally a scenario is created where after node reboot, application pods requesting devices appear before the device plugin pod
 		// exposing those devices as resource has restarted. The expected behavior is that the application pod fails at admission time.
-		framework.It("Keeps device plugin assignments across node reboots (no pod restart, no device plugin re-registration)", framework.WithFlaky(), func(ctx context.Context) {
+		// NOTE: this is flaky because all pods race to go running when kubelet restarts. So application pod will run with device plugin pod.
+		// It's possible, even though unlikely, that the device plugin pod manages to go running *and* the kubelet reports device healthy
+		// when the application pod enters admission. Hence the flakes.
```
Well, I don't think the device plugin should _ever_ re-register here because that can only happen if we delete the `triggerPathFile` and that doesn't happen after node reboot. @bart0sh mentioned [here](https://github.com/kubernetes/kubernetes/issues/128443#issuecomment-2466974720) that the test is consistently failing which aligns with the code I see here. Can you point to the test run where this test passed?

bart0sh replied:
@swatisehgal Here is my PR that "fixes" the test: https://github.com/kubernetes/kubernetes/pull/128995
You can see this test case passed in the pull-kubernetes-node-kubelet-containerd-flaky job

bart0sh replied:
@swatisehgal here is a test grid for the job: https://testgrid.k8s.io/sig-node-presubmits#pr-node-kubelet-serial-containerd-flaky&width=90
You can see at least 6 green test runs there. if you want to see more, just run `/test pull-kubernetes-node-kubelet-containerd-flaky` in [my PR](https://github.com/kubernetes/kubernetes/pull/128995).

swatisehgal replied:
> // NOTE: this is flaky because all pods race to go running when kubelet restarts. So application pod will run with device plugin pod.
		// It's possible, even though unlikely, that the device plugin pod manages to go running *and* the kubelet reports device healthy
		// when the application pod enters admission. Hence the flakes.

@ffromani I think we need to fix the comment here. Like I said before,  the device plugin pod should never go running in this case because the `triggerPathFile` wasn't removed and we should consistently see pod admission failure. 

I am also wondering if we should still mark this test flaky. With all the information we have after the investigation and the fix proposed here we should consistently see the pod failing at admission so makes me question is we should still consider it flaky.

swatisehgal replied:
> @swatisehgal here is a test grid for the job: https://testgrid.k8s.io/sig-node-presubmits#pr-node-kubelet-serial-containerd-flaky&width=90 You can see at least 6 green test runs there. if you want to see more, just run `/test pull-kubernetes-node-kubelet-containerd-flaky` in [my PR](https://github.com/kubernetes/kubernetes/pull/128995).

Thanks @bart0sh for this but I see that we have reached to an agreement about pursuing this PR and have closed https://github.com/kubernetes/kubernetes/pull/128995.

ffromani replied:
@swatisehgal I think you are right, thanks for pointing it out. The KEP-grade changes I was mentioning are for solving the general problem in production code, where we don't have the option to influence the sequence of events.

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -985,9 +988,9 @@ func testDevicePluginNodeReboot(f *framework.Framework, pluginSockDir string) {
-			err, _ = checkPodResourcesAssignment(v1PodResources, pod1.Namespace, pod1.Name, pod1.Spec.Containers[0].Name, SampleDeviceResourceName, []string{devID1})
-			framework.ExpectNoError(err, "inconsistent device assignment after node reboot")
-
+			// pods in terminal state will still be reported, see https://github.com/kubernetes/kubernetes/issues/119423
+			_, found := checkPodResourcesAssignment(v1PodResources, pod1.Namespace, pod1.Name, pod1.Spec.Containers[0].Name, SampleDeviceResourceName, []string{})
+			gomega.Expect(found).To(gomega.BeTrueBecause("%s/%s/%s failed admission, should not have devices registered", pod1.Namespace, pod1.Name, pod1.Spec.Containers[0].Name))
```
Here we are verifying that the pod has failed at admission and that it has no device allocated. This contradicts with the above comment that the pod may or may not fail at admission. Can you please clarify?

swatisehgal reviewed (commented):
Thanks for your deep dive on this topic and trying to capture the rationale in great detail. The overall direction seems okay to me but I have a couple of questions with the  proposed changes:

ffromani commented:
/test pull-kubernetes-node-kubelet-containerd-flaky
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -985,9 +988,17 @@ func testDevicePluginNodeReboot(f *framework.Framework, pluginSockDir string) {
+			// - sample device plugin pod, running and doing fine
+			// - our test pod, in failed state. Pods in terminal state will still be reported, see https://github.com/kubernetes/kubernetes/issues/119423
+			// so we care about our test pod, and it will be present in the returned list till 119423 is fixed, but since it failed admission it must not have
+			// any device allocated to it, hance we check for empty device set in the podresources response. So, we check that
```
Sorry for nitpicking:
```suggestion
			// any device allocated to it, hence we check for empty device set in the podresources response. So, we check that
```

ffromani replied:
thanks, fixed

swatisehgal commented on test/e2e_node/device_plugin_test.go:
```diff
@@ -985,9 +988,17 @@ func testDevicePluginNodeReboot(f *framework.Framework, pluginSockDir string) {
-			err, _ = checkPodResourcesAssignment(v1PodResources, pod1.Namespace, pod1.Name, pod1.Spec.Containers[0].Name, SampleDeviceResourceName, []string{devID1})
-			framework.ExpectNoError(err, "inconsistent device assignment after node reboot")
-
+			// if we got this far, podresources API will now report 2 entries:
+			// - sample device plugin pod, running and doing fine
+			// - our test pod, in failed state. Pods in terminal state will still be reported, see https://github.com/kubernetes/kubernetes/issues/119423
+			// so we care about our test pod, and it will be present in the returned list till 119423 is fixed, but since it failed admission it must not have
+			// any device allocated to it, hance we check for empty device set in the podresources response. So, we check that
+			// A. our test pod must be present in the list response *and*
+			// B. it has no devices assigned to it.
+			// anything else is unexpected and thus makes the test fail. Once 119423 is fixed, a better, simpler and more intuitive check will be for the
+			// test pod to not be present in the podresources list response, but till that time we're stuck with this approach.
+			_, found := checkPodResourcesAssignment(v1PodResources, pod1.Namespace, pod1.Name, pod1.Spec.Containers[0].Name, SampleDeviceResourceName, []string{})
+			gomega.Expect(found).To(gomega.BeTrueBecause("%s/%s/%s failed admission, should not have devices registered", pod1.Namespace, pod1.Name, pod1.Spec.Containers[0].Name))
```
Thanks for adding explanation here!

ffromani replied:
happy to help!

ffromani commented:
xref: https://github.com/kubernetes/kubernetes/pull/125918

ffromani commented:
/test pull-kubernetes-node-kubelet-containerd-flaky
/test pull-kubernetes-node-kubelet-serial-containerd
/test pull-kubernetes-node-kubelet-serial-containerd-sidecar-containers
/test pull-kubernetes-node-kubelet-serial-cpu-manager
/test pull-kubernetes-node-kubelet-serial-hugepages
/test pull-kubernetes-node-kubelet-serial-memory-manager
/test pull-kubernetes-node-kubelet-serial-topology-manager

ffromani commented:
/test pull-kubernetes-e2e-gce

kannon92 commented:
/priority important-soon
/triage accepted

swatisehgal commented:
Thanks @ffromani for addressing all the review comments. Looks good, the only pending item that requires some clarity is why we are still marking this test flaky. What are the scenarios where we would see unintended results?

ffromani commented:
> Thanks @ffromani for addressing all the review comments. Looks good, the only pending item that requires some clarity is why we are still marking this test flaky. What are the scenarios where we would see unintended results leading to test failure?

This makes sense, but the flaky label was added prior to this change. I wanted to address failures here :) Do we want to tackle the flakiness/non-flakiness situation in this PR?

swatisehgal commented:
> > Thanks @ffromani for addressing all the review comments. Looks good, the only pending item that requires some clarity is why we are still marking this test flaky. What are the scenarios where we would see unintended results leading to test failure?
> 
> This makes sense, but the flaky label was added prior to this change. I wanted to address failures here :) Do we want to tackle the flakiness/non-flakiness situation in this PR?

We had an offline discussion about this and agreed that it is worth understanding why this was made flaky ([Here ](https://github.com/kubernetes/kubernetes/pull/123386) is the PR that marked it flaky) and if we should keep it that way.

I think that this test should consistently pass now because we are creating a very controlled scenario where device plugin never re-registers after node-reboot resulting in the application pod **always** failing at admission (and without any device assignments). If that is not the case, I would like to understand why...

bart0sh commented:
/lgtm

/assign @SergeyKanzhelev @mrunalp

swatisehgal reviewed (commented):
/lgtm

We agreed in the SIG Node CI Meeting (22-01-2015) to monitor [node-kubelet-containerd-flaky](https://testgrid.k8s.io/sig-node-containerd#node-kubelet-containerd-flaky) job once this PR is merged to verify that it deflakes the test.

SergeyKanzhelev reviewed (approved):
/lgtm
/approve
